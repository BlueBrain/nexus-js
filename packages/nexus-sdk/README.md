<p align="center"><a href="https://github.com/BlueBrain/nexus-sdk-js" target="_blank" rel="noopener noreferrer"><img width="100" src="https://raw.githubusercontent.com/BlueBrain/nexus-webapp-commons/HEAD/nexus-js-logo.png" alt="Nexus Sdk logo"></a></p>

<h2 align="center">@bbp/nexus-sdk</h2>

<p align="center">
  <a href="https://codecov.io/gh/BlueBrain/nexus-sdk-js"><img src="https://codecov.io/gh/BlueBrain/nexus-sdk-js/branch/master/graph/badge.svg" alt="Coverage Status"></a>
  <a href="https://npmcharts.com/compare/@bbp/nexus-sdk?minimal=true"><img src="https://img.shields.io/npm/dm/@bbp/nexus-sdk.svg" alt="Downloads"></a>
  <a href="https://www.npmjs.com/package/@bbp/nexus-sdk"><img src="https://img.shields.io/npm/v/@bbp/nexus-sdk.svg" alt="Version"></a>
  <a href="https://www.npmjs.com/package/@bbp/nexus-sdk"><img src="https://img.shields.io/npm/l/@bbp/nexus-sdk.svg" alt="License"></a>
 <a href="https://twitter.com/intent/follow?screen_name=bluebrainnexus"><img alt="Follow on Twitter" src="https://img.shields.io/twitter/follow/bluebrainnexus.svg?style=social&label=Follow"></a>
</p>

<p align="center">
  <a href="https://www.epfl.ch/research/domains/bluebrain/">Blue Brain Project</a> |
  <a href="https://bluebrainnexus.io">Nexus</a> |
  <a href="https://bluebrainnexus.io/docs/">Nexus API Docs</a> |
  <a href="https://sandbox.bluebrainnexus.io">Sandbox</a>
</a>

<p align="center">
  <a href="https://codesandbox.io/search?refinementList%5Btags%5D%5B0%5D=bluebrainnexus&page=1&configure%5BhitsPerPage%5D=12">Examples</a> |
  <a href="#Get-Started">Getting Started</a> |
  <a href="#API">API Documentation</a> |
  <a href="#Development">Development</a> |
  <a href="#License">License</a>
</p>

<p align="center">This is a handy javascript toolkit that helps people just like you to interact with <a href="https://bluebrainnexus.io">Blue Brain Nexus</a> to build things like client-side applications, scripts, or backend services.
</p>

## Get Started

install with your favorite package manager

```sh
npm install @bbp/nexus-sdk
```

import inside your client or node.js

```typescript
import { createNexusClient } from '@bbp/nexus-sdk';

const nexus = createNexusClient({ uri: 'https://api.url' });
```

Now go and add use Nexus, such as adding [`resources`](./src/Resource#readme) to a [`project`](./src/Project#readme)

### Setup your token on client creation

```typescript
const nexus = createNexusClient({
  uri: 'https://api.url',
  token: 'my_bearer_token',
});
```

### Middleware

You can enhance the behaviour of Nexus Client with middlewares called Links.

```typescript
const nexus = createNexusClient({
  uri: 'https://api.url',
  links: [someMiddleware],
});
```

### Context

You can setup a "context" object that will be passed from links to links as part of the operation

```typescript
const myMiddleware: Link = (operation: Operation, forward: Link) => {
  const { myApiClient } = operation.context;
  // do something with your API client
  return forward(operation);
};

const nexus = createNexusClient({
  uri: 'https://api.url',
  context: {
    myApiClient,
  },
});
```

## Recipes

Set your bearer token before each request:

```typescript
const setToken: Link = (operation: Operation, forward: Link) => {
  const token = localStorage.getItem('myToken'); // get your token from somewhere
  const nextOperation = {
    ...operation,
    headers: {
      ...operation.headers,
      Authorization: `Bearer ${token}`,
    },
  };
  return forward(nextOperation);
};

const nexus = createNexusClient({ uri: 'https://api.url', links: [setToken] });
```

Log response time:

```typescript
const logResponseTime: Link = (operation: Operation, forward: Link) => {
  const time = Date.now();
  return forward(operation).map(data => {
    console.log('request took', Date.now() - time, 'ms');
    return data;
  });
};
```

Dispatch a redux action on every requests:

```typescript
const reduxDispatcher: (dispatch: any) => Link = dispatch => (
  operation: Operation,
  forward: Link,
) => {
  const prefix = '@@nexus';
  const actionName: string = getActionNameFromPath(operation.path);
  dispatch({
    name: `${prefix}/${actionName}_FETCHING`,
    payload: operation,
  });
  return forward(operation).map(data => {
    dispatch({ name: `${prefix}/${actionName}_SUCCESS`, payload: data });
    return data;
  });
};
```

### Node.js support

The Nexus SDK relies on [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), so in order to use this library in Node.js, you need to provide a `fetch` implementation when creating a new client. We recommend using [`node-fetch`](https://www.npmjs.com/package/node-fetch)

Request Cancellation is using `AbortController`, so you need to polyfill it. As [documented](https://www.npmjs.com/package/node-fetch#request-cancellation-with-abortsignal) in `node-fetch`, you can use [`abort-controller`](https://www.npmjs.com/package/abort-controller) as a polyfill.

Example:

```javascript
const fetch = require('node-fetch');
require('abort-controller/polyfill');

const nexus = createNexusClient({
  uri: 'https://sandbox.bluebrainnexus.io',
  fetch,
});

nexus.Organization.list()
  .then(orgs => console.log(orgs))
  .catch(e => console.error(e));
```

## API

#### Admin

[Orgs](./src/Organization) | [Projects](./src/Project)

#### Knowledge Graph

[Resources](./src/Resource) | [Schema](./src/Schema) | [Files](./src/File) | [Views](./src/View) | [Storage](./src/Storage) | [Resolver](./src/Resolver)

#### Identity and Access Management

[Realm](./src/Realm) | [Permissions](./src/Permissions) | [Identities](./src/Identity) | [ACL](./src/ACL)

#### Misc.

The client also exposes bare http methods that can be use for operations on a resources using the `_self` url for example.

```typescript
/**
 * Methods exposed are:
 * - nexus.httpGet
 * - nexus.httpPost
 * - nexus.httpPut
 * - nexus.httpDelete
 * - nexus.poll
 */

// post something as text
nexus.httpPost({
  path: 'https://mySelfUrl.com',
  headers: { 'Content-type': 'text/plain' },
  body: 'Some text',
});

// get something as blob
nexus.get({ path: 'https://mySelfUrl.com', context: { as: 'blob' } });

// poll something
nexus.poll({ path: 'https://mySelfUrl.com', context: { pollTime: 1000 } });
```

You can also import constants for defautl view IDs or default Schema IDs:

```typescript
import { DEFAULT_ELASTIC_SEARCH_VIEW_ID, DEFAULT_SPARQL_VIEW_ID, DEFAULT_SCHEMA_ID } from '@bbp/nexus-sdk';

nexus.View.get('myorg', 'myproject', DEFAULT_ELASTIC_SEARCH_VIEW_ID).then(view => { // do something with my view});
```

## Development

### How is the repo structured?

This repo is managed as a _monorepo_ using [lerna](https://github.com/lerna/lerna) that is composed of multiple npm packages.

> Make sure you perform the make actions in the repository root directory!

### Using Docker

> If you don't have Node.js installed on your machine, you can run a "docker shell" with `make dshell` from which you'll have a fully working Node.js environment.
> Make sure you have already installed both [Docker Engine](https://docs.docker.com/install/) and [Docker Compose](https://docs.docker.com/compose/install/).

### Do things

- Install: `make install`
- Build: `make build`
- Test: `make test`
- Lint: `make lint`

# License

[Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0)
