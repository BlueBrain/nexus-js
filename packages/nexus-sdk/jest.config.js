const base = require('../../jest.config.base.js');

module.exports = {
  ...base,
  name: '@bbp/nexus-sdk',
  displayName: '@bbp/nexus-sdk',
};
