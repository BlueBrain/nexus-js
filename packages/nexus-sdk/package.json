{
  "name": "@bbp/nexus-sdk",
  "version": "1.1.5",
  "description": "REST API abstraction for Nexus",
  "keywords": [
    "rest",
    "js",
    "node",
    "sdk",
    "typescript"
  ],
  "author": "Julien Machon <julienmachon@gmail.com>",
  "homepage": "https://github.com/BlueBrain/nexus-webapp-commons#readme",
  "license": "Apache-2.0",
  "main": "./lib/index.js",
  "module": "./es/index.js",
  "unpkg": "./dist/index.js",
  "typings": "./lib/index.d.ts",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/BlueBrain/nexus-webapp-commons.git"
  },
  "scripts": {
    "prebuild": "npm run clean",
    "build": "rollup -c",
    "build:watch": "rollup -c -w",
    "clean": "rm -fr es/ && rm -fr dist/ && rm -fr lib/"
  },
  "bugs": {
    "url": "https://github.com/BlueBrain/nexus-webapp-commons/issues"
  },
  "dependencies": {
    "@bbp/nexus-link": "^1.1.5"
  },
  "publishConfig": {
    "access": "public"
  }
}
