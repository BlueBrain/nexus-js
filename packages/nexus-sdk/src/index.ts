export * from './nexusSdk';
export * from './types';
