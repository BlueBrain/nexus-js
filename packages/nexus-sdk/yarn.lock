# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


decode-uri-component@^0.2.0:
  version "0.2.0"
  resolved "https://registry.yarnpkg.com/decode-uri-component/-/decode-uri-component-0.2.0.tgz#eb3913333458775cb84cd1a1fae062106bb87545"
  integrity sha1-6zkTMzRYd1y4TNGh+uBiEGu4dUU=

query-string@^6.9.0:
  version "6.9.0"
  resolved "https://registry.yarnpkg.com/query-string/-/query-string-6.9.0.tgz#1c3b727c370cf00f177c99f328fda2108f8fa3dd"
  integrity sha512-KG4bhCFYapExLsUHrFt+kQVEegF2agm4cpF/VNc6pZVthIfCc/GK8t8VyNIE3nyXG9DK3Tf2EGkxjR6/uRdYsA==
  dependencies:
    decode-uri-component "^0.2.0"
    split-on-first "^1.0.0"
    strict-uri-encode "^2.0.0"

split-on-first@^1.0.0:
  version "1.1.0"
  resolved "https://registry.yarnpkg.com/split-on-first/-/split-on-first-1.1.0.tgz#f610afeee3b12bce1d0c30425e76398b78249a5f"
  integrity sha512-43ZssAJaMusuKWL8sKUBQXHWOpq8d6CfN/u1p4gUzfJkM05C8rxTmYrkIPTXapZpORA6LkkzcUulJ8FqA7Uudw==

strict-uri-encode@^2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/strict-uri-encode/-/strict-uri-encode-2.0.0.tgz#b9c7330c7042862f6b142dc274bbcc5866ce3546"
  integrity sha1-ucczDHBChi9rFC3CdLvMWGbONUY=
