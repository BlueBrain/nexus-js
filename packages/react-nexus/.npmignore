src
tsconfig.json
rollup.config.js
jest.config.js