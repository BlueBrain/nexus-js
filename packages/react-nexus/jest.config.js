const base = require('../../jest.config.base.js');

module.exports = {
  ...base,
  name: '@bbp/react-nexus',
  displayName: '@bbp/react-nexus',
};
