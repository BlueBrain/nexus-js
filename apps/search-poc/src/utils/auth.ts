import { Realm } from '@bbp/nexus-sdk';
import { UserManager, User } from 'oidc-client';
import { Operation, Link, Observable } from '@bbp/nexus-link';
import { SETTINGS } from '../config';

// creates a OIDC config based on input realm
export function getConfig(realm?: Realm) {
  return {
    authority: realm ? realm._issuer : '',
    client_id: SETTINGS.clientId,
    redirect_uri: SETTINGS.redirectUrl,
    popup_redirect_uri: SETTINGS.redirectUrl,
    post_logout_redirect_uri: SETTINGS.redirectUrl,
    response_type: 'id_token token',
    loadUserInfo: true,
  };
}

// link that sets the bearer token before each nexus call
export const setToken: Link = (operation: Operation, forward?: Link) => {
  const token = localStorage.getItem(SETTINGS.bearerTokenKey);
  const nextHeaders: any = { ...operation.headers };
  token && (nextHeaders['Authorization'] = `Bearer ${token}`);

  const nextOperation = {
    ...operation,
    headers: nextHeaders,
  };

  return forward ? forward(nextOperation) : new Observable();
};

/**
 * This function will:
 * - try to create a user manager instance if the realms settings are available from local storage
 * - set event listeners in order to keep the bearer token in sync in local storage
 * - look for new session on creation
 * - load user info
 * - return a ref of user manager instance and the user
 */
export async function setUpSession(): Promise<[UserManager, User | null]> {
  let user: User | null = null;
  let userManager: UserManager | null = null;
  // Get preferred realm
  const preferredRealm: Realm = JSON.parse(
    localStorage.getItem(SETTINGS.preferredRealmKey) || '{}',
  );

  // then create User Manager with preferred realm issuer
  // or empty user manager if no preferred realm
  userManager = new UserManager(getConfig(preferredRealm));

  // maybe we're getting new user session after redirect
  try {
    user = await userManager.signinRedirectCallback();
  } catch (error) {}
  // if not, check if we already had a running session
  if (!user) {
    try {
      user = await userManager.getUser();
    } catch (error) {
      // if not, that's fine
    }
  }

  // Set events
  userManager.events.addUserLoaded((user: User) => {
    localStorage.setItem(SETTINGS.bearerTokenKey, user.id_token);
  });

  userManager.events.addAccessTokenExpired(() => {
    localStorage.removeItem(SETTINGS.bearerTokenKey);
  });

  return [userManager, user];
}
