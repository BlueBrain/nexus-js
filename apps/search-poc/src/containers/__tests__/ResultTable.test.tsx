import ResultTableContainer from '../ResultTable';
import React from "react";
import { mount } from "enzyme";
import { BrowserRouter as Router } from 'react-router-dom';
import { NexusProvider } from '@bbp/react-nexus';
import { NexusClient, createNexusClient } from '@bbp/nexus-sdk';
import {  setToken } from '../../utils/auth.ts';

describe('ResultTableContainer', () => {
    
    it('ResultTableContainer renders correctly', () => {
        const nexus: NexusClient = createNexusClient({
            fetch,
            uri: 'test',
            links: [setToken],
          });
        const props = {
            dataQuery: 'a query',
            orgLabel: 'org',
            projectLabel: 'project Label',
            viewId: 'viewId',
            handleRowClick: jest.fn
          };
        const result = mount(<NexusProvider nexusClient={nexus}><Router><ResultTableContainer {...props} /></Router></NexusProvider>)
        expect(result).toMatchSnapshot();
    });

});

