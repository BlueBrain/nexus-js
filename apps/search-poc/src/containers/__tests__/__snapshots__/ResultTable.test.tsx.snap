// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultTableContainer ResultTableContainer renders correctly 1`] = `
<NexusProvider
  nexusClient={
    Object {
      "ACL": Object {
        "append": [Function],
        "create": [Function],
        "delete": [Function],
        "list": [Function],
        "poll": [Function],
        "replace": [Function],
        "subtract": [Function],
      },
      "File": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "link": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Identity": Object {
        "list": [Function],
        "poll": [Function],
      },
      "Organization": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Permissions": Object {
        "append": [Function],
        "delete": [Function],
        "list": [Function],
        "poll": [Function],
        "replace": [Function],
        "subtract": [Function],
      },
      "Project": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Realm": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Resolver": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Resource": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "links": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Schema": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Storage": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "View": Object {
        "create": [Function],
        "deprecate": [Function],
        "elasticSearchQuery": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "pollStatistics": [Function],
        "sparqlQuery": [Function],
        "statistics": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "httpDelete": [Function],
      "httpGet": [Function],
      "httpPatch": [Function],
      "httpPost": [Function],
      "httpPut": [Function],
      "poll": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <ResultTableContainer
        dataQuery="a query"
        handleRowClick={[Function]}
        orgLabel="org"
        projectLabel="project Label"
        viewId="viewId"
      >
        <Spin
          size="default"
          spinning={true}
          wrapperClassName=""
        >
          <div
            className="ant-spin ant-spin-spinning"
          >
            <span
              className="ant-spin-dot ant-spin-dot-spin"
            >
              <i
                className="ant-spin-dot-item"
              />
              <i
                className="ant-spin-dot-item"
              />
              <i
                className="ant-spin-dot-item"
              />
              <i
                className="ant-spin-dot-item"
              />
            </span>
          </div>
        </Spin>
      </ResultTableContainer>
    </Router>
  </BrowserRouter>
</NexusProvider>
`;
